cli_tools/diagnostics/main_windows.go

```diff
@@ -0,0 +1,241 @@
+			path, err := command.run()
+			if err != nil {
+				log.Printf("Error: %s while running %v", err, command)
+			} else {
+				ch <- path
+			}
+		}(command)
+	}
+	wg.Wait()
+	close(ch)
+	for path := range ch {
+		paths = append(paths, path)
+	}
+	return paths
+}
+
+func gatherSystemLogs(logs chan logFolder) {
+	var commands = []runner{
+		cmd{"systeminfo.exe", "", "systeminfo.txt", false},
+		cmd{"bcdedit.exe", "", "bcdedit.txt", false},
+		cmd{"sc.exe", "query type=driver", "drivers.txt", false},
+		cmd{"pnputil.exe", "/e", "pnputil.txt", false},
+		cmd{"msinfo32.exe", "/report msinfo32.txt", "msinfo32.txt", true},
+		wmiQuery{"Win32_UserAccount", `root\CIMv2`, "users.txt"},
+	}
+
+	logs <- logFolder{"System", runAll(commands)}
+}
+
+func gatherDiskLogs(logs chan logFolder) {
+	var commands = []runner{
+		wmiQuery{"MSFT_Disk", `root\Microsoft\Windows\Storage`, "disks.txt"},
+		wmiQuery{"MSFT_Volume", `root\Microsoft\Windows\Storage`, "volumes.txt"},
+		wmiQuery{"MSFT_Partition", `root\Microsoft\Windows\Storage`, "partitions.txt"},
+	}
+
+	logs <- logFolder{"Disk", runAll(commands)}
+}
+
+func gatherNetworkLogs(logs chan logFolder) {
+	var commands = []runner{
+		cmd{"nslookup.exe", "8.8.8.8", "nslookup_dns.txt", false},
+		cmd{"tracert.exe", "www.gstatic.com", "tracert_gstatic.txt", false},
+		cmd{"ping.exe", "-n 10 8.8.8.8", "ping_dns.txt", false},
+		cmd{"ping.exe", "-n 10 www.gstatic.com", "ping_gstatic.txt", false},
+		cmd{"ipconfig.exe", "/all", "ipconfig.txt", false},
+		cmd{"route.exe", "print", "route.txt", false},
+		cmd{"netstat.exe", "-anb", "netstat.txt", false},
+		wmiQuery{"MSFT_NetFirewallRule", `root\StandardCimv2`, "firewall.txt"},
+	}
+
+	logs <- logFolder{"Network", runAll(commands)}
+}
+
+func gatherProgramLogs(logs chan logFolder) {
+	var commands = []runner{
+		wmiQuery{"Win32_Process", `root\Cimv2`, "processes.txt"},
+		wmiQuery{"Win32_Service", `root\Cimv2`, "services.txt"},
+		wmiQuery{"MSFT_ScheduledTask", `root\Microsoft\Windows\TaskScheduler`, "scheduled_tasks.txt"},
+	}
+
+	logs <- logFolder{"Program", runAll(commands)}
+}
+
+func gatherEventLogs(logs chan logFolder) {
+	dirs := []string{`C:\Windows\System32\winevt\Logs`}
+	paths := make([]string, 0)
+	// Recursively gather all the log files in a directory
+	for len(dirs) > 0 {
+		dir := dirs[0]
+		dirs = dirs[1:]
+		files, _ := ioutil.ReadDir(dir)
+		for _, f := range files {
+			path := filepath.Join(dir, f.Name())
+			if f.IsDir() {
+				dirs = append(dirs, path)
+			} else {
+				paths = append(paths, path)
+			}
+		}
+	}
+	logs <- logFolder{"Event", paths}
+}
+
+func gatherTraceLogs(logs chan logFolder) {
+	traceStart := cmd{"wpr.exe", "-start CPU -start DiskIO -start FileIO -start Network", "trace.etl", true}
+	traceStop := cmd{"wpr.exe", "-stop trace.etl", "trace.etl", true}
+
+	traceStart.run()
+	time.Sleep(10 * time.Minute)
+	paths := runAll([]runner{
+		traceStop,
+	})
+	logs <- logFolder{"Trace", paths}
+
+}
+
+func gatherLogs(trace bool) []logFolder {
```

Comment:
This could be simplified into something like this:

runFuncs := []func(logs chan logFolder) {
  gatherSystemLogs
  gatherDiskLogs
  ...
}
if trace {
  runFuncs = append(runFuncs, gatherTraceLogs)
}
folderCount := len(runFuncs)

for _, run := range runFuncs {
  go run(ch)
}

Reply:
Done